java/json/src/main/java/org/bondlib/SimpleJsonReader.java

```diff
@@ -0,0 +1,480 @@
+        // nothing to do here
+    }
+
+    @Override
+    public final void readContainerBegin() throws IOException {
+        // skip a START_ARRAY token that is expected here
+        final JsonToken currentToken = this.advanceParserToNextToken();
+        if (currentToken != JsonToken.START_ARRAY) {
+            throw createInvalidBondDataException(
+                    "Current JSON token is '%s' where '%s' is expected.",
+                    currentToken,
+                    JsonToken.START_ARRAY);
+        }
+    }
+
+    @Override
+    public final void readContainerEnd() throws IOException {
+        // need to clear the peek flag to make sure nested collections work properly
+        //
+        // note that this amounts to skipping a END_ARRAY token which is expected to
+        // be the current (peeked) token when this method is entered
+        this.hasPeekedParserForNextToken = false;
+    }
+
+    @Override
+    public final boolean readContainerItemBegin() throws IOException {
+        // peek the next token to see whether it is end of an array
+        //
+        // note that this is the only place where peekParserForNextToken is called and since clients
+        // of TextProtocolReader should call either readXXX (some value) or readContainerEnd between
+        // any two calls calls to readContainerItemBegin, the hasPeekedParserForNextToken flag is
+        // expected to be clear every time this method is called
+        //
+        // also note that unless we have reached the end on a container, the caller will invoke
+        // one of the methods to read a value, which will call method advanceParserToNextToken,
+        // which will check whether the token is null, and thus EOF will be detected when reading
+        // a value
+        final JsonToken nextToken = this.peekParserForNextToken();
+        return nextToken != JsonToken.END_ARRAY;
+    }
+
+    @Override
+    public final void readContainerItemEnd() throws IOException {
+        // nothing to do here
+    }
+
+    @Override
+    public final byte readInt8() throws IOException {
+        final JsonToken currentToken = this.advanceParserToNextToken();
+        if (currentToken != JsonToken.VALUE_NUMBER_INT) {
+            throw createInvalidBondDataException(
+                    "Current JSON token is '%s' where '%s' for Bond 'int8' data type is expected.",
+                    currentToken,
+                    JsonToken.VALUE_NUMBER_INT);
+        }
+        // an int8 value must be represented by INT (Jackson's minimal "optimal type") and be within bounds
+        final JsonParser.NumberType numberType = this.parser.getNumberType();
+        if (numberType == JsonParser.NumberType.INT) {
+            final int valueAsInt = this.parser.getIntValue();
+            if (valueAsInt >= Byte.MIN_VALUE && valueAsInt <= Byte.MAX_VALUE) {
+                return (byte) valueAsInt;
+            }
+        }
+        throw createInvalidBondDataException(
```

Comment:
I would find this easier to read in the else branch of the range check than as a fall-through. If other values of `numberType` are float/double or something, I don't think it accurately describes the error when that check fails, either.

Reply:
It is intended as fall-through. If numberType is long, bigint, or an int outside the range, it should all be an error. Regarding float/double, that would be caught in the first check since the token type would not be VALUE_NUMBER_INT.

---
In reply to: [143860436](https://github.com/Microsoft/bond/pull/632#discussion_r143860436) [](ancestors = 143860436)